File: backend/dotnet/ClinicAppointments.Api/Queries/IdentityQueries.cs
using ClinicAppointments.Db;
using ClinicAppointments.Db.Entities;
using Microsoft.EntityFrameworkCore;

namespace ClinicAppointments.Api.Queries;

static class IdentityQueries
{
    public static async Task<Identity?> FindIdentity(this AppDbContext ctx, string? email = null, Guid? id = null)
    {
        var admins = ctx.Admins.AsNoTracking()
            .ByEmail(email)
            .ById(id)
            .Select(it => new Identity(it.Id, it.Name, it.Email, it.Password, "admin"));
        if (await admins.FirstOrDefaultAsync() is Identity result) return result;
        var secretaries = ctx.Secretaries
            .AsNoTracking()
            .ByEmail(email)
            .ById(id)
            .Select(it => new Identity(it.Id, it.Name, it.Email, it.Password, "secretary"));
        return await secretaries.FirstOrDefaultAsync();
    }

    public static IQueryable<Admin> ByEmail(this IQueryable<Admin> admins, string? email) =>
        email is null ? admins : admins.Where(it => it.Email == email);

    public static IQueryable<Admin> ById(this IQueryable<Admin> admins, Guid? id) =>
        id is null ? admins : admins.Where(it => it.Id == id);
}

public record Identity(Guid Id, string Name, string Email, string Password, string Role);

File: db-types/src/db/queries/identity_queries.ts
/** Types generated for queries found in "generated/queries/identity_queries.sql" */
import { PreparedQuery } from '@pgtyped/runtime';

/** 'GetIdentityByEmail' parameters type */
export interface IGetIdentityByEmailParams {
  email?: string | null | void;
}

/** 'GetIdentityByEmail' return type */
export interface IGetIdentityByEmailResult {
  email: string | null;
  id: string | null;
  name: string | null;
  password: string | null;
  role: string | null;
}

/** 'GetIdentityByEmail' query type */
export interface IGetIdentityByEmailQuery {
  params: IGetIdentityByEmailParams;
  result: IGetIdentityByEmailResult;
}

const getIdentityByEmailIR: any = {"usedParamSet":{"email":true},"params":[{"name":"email","required":false,"transform":{"type":"scalar"},"locs":[{"a":99,"b":104},{"a":225,"b":230}]}],"statement":"SELECT \"id\", \"name\", \"email\", \"password\", 'admin' as \"role\"\nFROM \"admins\"\nWHERE \"admins\".\"email\" = :email\nUNION\nSELECT \"id\", \"name\", \"email\", \"password\", 'secretary' as \"role\"\nFROM \"secretaries\"\nWHERE \"secretaries\".\"email\" = :email\nLIMIT 1"};

/**
 * Query generated from SQL:
 * ```
 * SELECT "id", "name", "email", "password", 'admin' as "role"
 * FROM "admins"
 * WHERE "admins"."email" = :email
 * UNION
 * SELECT "id", "name", "email", "password", 'secretary' as "role"
 * FROM "secretaries"
 * WHERE "secretaries"."email" = :email
 * LIMIT 1
 * ```
 */
export const getIdentityByEmail = new PreparedQuery<IGetIdentityByEmailParams,IGetIdentityByEmailResult>(getIdentityByEmailIR);


/** 'GetIdentityById' parameters type */
export interface IGetIdentityByIdParams {
  id?: string | null | void;
}

/** 'GetIdentityById' return type */
export interface IGetIdentityByIdResult {
  email: string | null;
  id: string | null;
  name: string | null;
  password: string | null;
  role: string | null;
}

/** 'GetIdentityById' query type */
export interface IGetIdentityByIdQuery {
  params: IGetIdentityByIdParams;
  result: IGetIdentityByIdResult;
}

const getIdentityByIdIR: any = {"usedParamSet":{"id":true},"params":[{"name":"id","required":false,"transform":{"type":"scalar"},"locs":[{"a":96,"b":98},{"a":216,"b":218}]}],"statement":"SELECT \"id\", \"name\", \"email\", \"password\", 'admin' as \"role\"\nFROM \"admins\"\nWHERE \"admins\".\"id\" = :id\nUNION\nSELECT \"id\", \"name\", \"email\", \"password\", 'secretary' as \"role\"\nFROM \"secretaries\"\nWHERE \"secretaries\".\"id\" = :id\nLIMIT 1"};

/**
 * Query generated from SQL:
 * ```
 * SELECT "id", "name", "email", "password", 'admin' as "role"
 * FROM "admins"
 * WHERE "admins"."id" = :id
 * UNION
 * SELECT "id", "name", "email", "password", 'secretary' as "role"
 * FROM "secretaries"
 * WHERE "secretaries"."id" = :id
 * LIMIT 1
 * ```
 */
export const getIdentityById = new PreparedQuery<IGetIdentityByIdParams,IGetIdentityByIdResult>(getIdentityByIdIR);


